File: 프로그래머스/lv2/권채린/구명보트.cpp
#include <string>
#include <vector>
#include <algorithm>
// 최대 2명이 탈 수 있으며, 몸무게 최댓값 보다는 무조건 크다 => 가장 무거운 사람을 태우고, 가장 가벼운 사람을 추가로 태울 수 있을 지 체크
using namespace std;

int solution(vector<int> people, int limit) {
    int answer = 0;
    sort(people.begin(), people.end());
  
    int start=0;
    int end=people.size()-1;
  
    while(start<=end) {
        if(people[start]+people[end]<=limit)
            start++;
        end--;
        answer++;
    }

    return answer;
}

File: 프로그래머스/lv2/이하은/구명보트.java
import java.util.*;

/*
O(N), O(NlogN) N <= 5만

모든 사람을 구출하기 위해 필요한 구명보트 개수의 최솟값
최대 2명까지만 탑승 가능 

무게가 적은 + 많은 사람 같이 묶어서 태우기 
*/
class Solution {
    public int solution(int[] arr, int limit) {
        int cnt = 0;
        Arrays.sort(arr);
        
        // 50 50 70 80 
        // 50 70 80 
        int l = 0, r = arr.length - 1;
        while(l <= r){
            if(arr[l] + arr[r] <= limit) l++; 
            r--;
            cnt++;
        }
        
        return cnt;
    }
}